File: src/harness/core/agent_loop.py
"""
Agent Loop - The core execution engine.

Implements the ReAct-style loop that drives agent behavior.
"""

import os
import time
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from harness.llm.base import LLMClient, ToolDefinition
from harness.memory.context_builder import ContextBuilder
from harness.memory.session import SessionManager
from harness.tools.executor import ToolContext, ToolExecutor
from harness.types import (
    LoopResult,
    LoopState,
    Message,
    ProgressCallback,
    ProgressEvent,
    ProgressEventType,
    Session,
    ToolCall,
)


@dataclass
class LoopConfig:
    """Configuration for agent loop."""
    max_iterations: int = 100
    timeout_per_tool: float = 30.0
    enable_parallel_tools: bool = True
    retry_on_error: int = 3
    enable_progress: bool = True  # Enable progress events


class AgentLoop:
    """
    The core agent execution loop.

    Implements a ReAct-style loop:
    1. Build context from session
    2. Call LLM
    3. Execute tools if needed
    4. Repeat until done
    """

    def __init__(
        self,
        llm_client: LLMClient,
        tool_executor: ToolExecutor,
        context_builder: ContextBuilder,
        session_manager: SessionManager,
        config: LoopConfig | None = None,
    ):
        self.llm = llm_client
        self.tools = tool_executor
        self.context = context_builder
        self.sessions = session_manager
        self.config = config or LoopConfig()

        self.state = LoopState.IDLE
        self._interrupt_flag = False
        self._on_progress: ProgressCallback | None = None
        self._loop_start_time: float | None = None
        self._last_event_time: float | None = None

    def _emit_progress(
        self,
        event_type: ProgressEventType,
        message: str,
        data: dict | None = None,
        duration_ms: float | None = None,
    ) -> None:
        """Emit a progress event if callback is set."""
        if self._on_progress and self.config.enable_progress:
            event = ProgressEvent(
                type=event_type,
                message=message,
                data=data or {},
                duration_ms=duration_ms,
            )
            self._on_progress(event)
            self._last_event_time = time.time()

    async def run(
        self,
        prompt: str,
        session: Session,
        tools: list[ToolDefinition] | None = None,
        on_chunk: Callable[[str], None] | None = None,
        on_progress: ProgressCallback | None = None,
    ) -> LoopResult:
        """
        Run the agent loop.

        Args:
            prompt: User input
            session: Current session
            tools: Available tools
            on_chunk: Streaming callback
            on_progress: Progress event callback

        Returns:
            LoopResult: Final result
        """
        self._on_progress = on_progress
        self._loop_start_time = time.time()
        self._interrupt_flag = False

        # Emit loop start
        prompt_preview = prompt[:100] + "..." if len(prompt) > 100 else prompt
        self._emit_progress(
            ProgressEventType.LOOP_START,
            f"Starting agent loop",
            {"prompt": prompt_preview, "session_id": session.id},
        )

        self.state = LoopState.BUILDING_CONTEXT
        iteration = 0
        total_usage = session.token_usage

        try:
            while iteration < self.config.max_iterations:
                # Emit iteration event
                self._emit_progress(
                    ProgressEventType.ITERATION,
                    f"Iteration {iteration + 1}/{self.config.max_iterations}",
                    {"iteration": iteration + 1},
                )

                # Check for interruption
                if self._interrupt_flag:
                    self.state = LoopState.INTERRUPTED
                    self._emit_progress(
                        ProgressEventType.LOOP_END,
                        "Loop interrupted",
                        {"status": "interrupted", "iterations": iteration},
                    )
                    return LoopResult(
                        status=LoopState.INTERRUPTED,
                        session=session,
                        iterations=iteration,
                    )

                # Build context
                self.state = LoopState.BUILDING_CONTEXT
                context_build_start = time.time()
                self._emit_progress(
                    ProgressEventType.STATE_CHANGE,
                    "Building context",
                    {"state": LoopState.BUILDING_CONTEXT.value},
                )
                context = self.context.build(session, prompt if iteration == 0 else None)
                context_duration = (time.time() - context_build_start) * 1000

                # Call LLM
                self.state = LoopState.CALLING_LLM
                llm_call_start = time.time()
                self._emit_progress(
                    ProgressEventType.LLM_CALL,
                    f"Calling LLM: {self.llm.model_name}",
                    {"model": self.llm.model_name, "message_count": len(context.messages)},
                )

                response = await self.llm.call(
                    messages=context.messages,
                    tools=tools,
                    system=context.system_prompt,
                )

                llm_duration = (time.time() - llm_call_start) * 1000
                self._emit_progress(
                    ProgressEventType.LLM_RESPONSE,
                    f"LLM responded",
                    {
                        "stop_reason": response.stop_reason.value,
                        "input_tokens": response.usage.input_tokens,
                        "output_tokens": response.usage.output_tokens,
                    },
                    duration_ms=llm_duration,
                )

                # Update usage
                total_usage.input_tokens += response.usage.input_tokens
                total_usage.output_tokens += response.usage.output_tokens

                # Add assistant message
                assistant_msg = Message(
                    role="assistant",
                    content=response.content,
                )
                session.add_message(assistant_msg)

                # Check if we need tools
                if response.is_tool_use:
                    self.state = LoopState.EXECUTING_TOOLS
                    self._emit_progress(
                        ProgressEventType.STATE_CHANGE,
                        f"Executing {len(response.tool_calls)} tool(s)",
                        {
                            "state": LoopState.EXECUTING_TOOLS.value,
                            "tool_count": len(response.tool_calls),
                            "tools": [tc.name for tc in response.tool_calls],
                        },
                    )

                    # Execute tools
                    tool_results = await self._execute_tools(
                        response.tool_calls,
                        session,
                    )

                    # Add tool results to session
                    for result in tool_results:
                        tool_msg = Message(
                            role="tool",
                            content=result.content,
                            metadata={"tool_call_id": result.tool_call_id},
                        )
                        session.add_message(tool_msg)

                    iteration += 1
                    continue

                # Done!
                self.state = LoopState.COMPLETED
                session.token_usage = total_usage

                # Emit completion
                total_duration = (time.time() - self._loop_start_time) * 1000
                self._emit_progress(
                    ProgressEventType.LOOP_END,
                    "Loop completed successfully",
                    {
                        "status": "completed",
                        "iterations": iteration + 1,
                        "total_tokens": total_usage.total_tokens,
                    },
                    duration_ms=total_duration,
                )

                return LoopResult(
                    status=LoopState.COMPLETED,
                    session=session,
                    messages=session.messages,
                    final_response=response.content,
                    iterations=iteration,
                    token_usage=total_usage,
                )

            # Max iterations reached
            self.state = LoopState.ERROR
            self._emit_progress(
                ProgressEventType.ERROR,
                "Max iterations reached",
                {"iterations": iteration},
            )
            return LoopResult(
                status=LoopState.ERROR,
                session=session,
                messages=session.messages,
                iterations=iteration,
                error="Max iterations reached",
                token_usage=total_usage,
            )

        except Exception as e:
            self.state = LoopState.ERROR
            self._emit_progress(
                ProgressEventType.ERROR,
                f"Error: {str(e)}",
                {"error": str(e), "type": type(e).__name__},
            )
            return LoopResult(
                status=LoopState.ERROR,
                session=session,
                messages=session.messages,
                iterations=iteration,
                error=str(e),
                token_usage=total_usage,
            )

    async def _execute_tools(
        self,
        tool_calls: list[ToolCall],
        session: Session,
    ) -> list:
        """Execute tool calls with progress tracking."""
        from harness.tools.permissions import PermissionSet

        context = ToolContext(
            session_id=session.id,
            working_directory=Path(os.getcwd()),
            permissions=PermissionSet.sandbox(os.getcwd()),
        )

        results = []
        for tool_call in tool_calls:
            # Emit tool call start
            self._emit_progress(
                ProgressEventType.TOOL_CALL,
                f"Executing: {tool_call.name}",
                {
                    "tool": tool_call.name,
                    "tool_call_id": tool_call.id,
                    "arguments": tool_call.arguments,
                },
            )

            tool_start = time.time()
            result = await self.tools.execute(tool_call, context)
            tool_duration = (time.time() - tool_start) * 1000

            # Emit tool result
            status = "success" if result.success else "failed"
            self._emit_progress(
                ProgressEventType.TOOL_RESULT,
                f"Tool {tool_call.name}: {status}",
                {
                    "tool": tool_call.name,
                    "tool_call_id": tool_call.id,
                    "success": result.success,
                    "error": result.error if not result.success else None,
                },
                duration_ms=tool_duration,
            )

            results.append(result)

        return results

    def interrupt(self) -> None:
        """Interrupt the current loop."""
        self._interrupt_flag = True

File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/core/AgentLoop.java
package com.harness.core;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.random.RandomGenerator;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.harness.types.*;

/**
 * Agent loop - the core execution engine.
 *
 * Implements a ReAct-style loop:
 * 1. Build context from session
 * 2. Call LLM
 * 3. Execute tools if needed
 * 4. Repeat until done
 *
 * Robustness features:
 * - LLM retry with exponential backoff and jitter
 * - Tool execution timeout protection
 * - Configurable retry count
 * - Intelligent error handling with ErrorHandler
 */
public class AgentLoop {

    private static final Logger logger = LoggerFactory.getLogger(AgentLoop.class);
    private static final RandomGenerator random = RandomGenerator.getDefault();

    private final LLMClient llmClient;
    private final ToolExecutor toolExecutor;
    private final ContextBuilder contextBuilder;
    private final LoopConfig config;
    private final ErrorHandler errorHandler;

    private volatile LoopState state = LoopState.IDLE;
    private final AtomicBoolean interrupted = new AtomicBoolean(false);
    private final AtomicInteger iteration = new AtomicInteger(0);

    public AgentLoop(LLMClient llmClient, ToolExecutor toolExecutor, LoopConfig config) {
        this.llmClient = llmClient;
        this.toolExecutor = toolExecutor;
        this.config = config;
        this.contextBuilder = new ContextBuilder();
        this.errorHandler = new ErrorHandler();
    }

    public AgentLoop(LLMClient llmClient, ToolExecutor toolExecutor) {
        this(llmClient, toolExecutor, LoopConfig.defaults());
    }

    /**
     * Run the agent loop synchronously.
     */
    public LoopResult run(String prompt, Session session) {
        return runAsync(prompt, session).join();
    }

    /**
     * Run the agent loop with default session.
     */
    public LoopResult run(String prompt) {
        return run(prompt, Session.create());
    }

    /**
     * Run the agent loop asynchronously.
     */
    public CompletableFuture<LoopResult> runAsync(String prompt, Session session) {
        return CompletableFuture.supplyAsync(() -> executeLoop(prompt, session));
    }

    /**
     * Main loop execution.
     */
    private LoopResult executeLoop(String prompt, Session session) {
        logger.info("Starting agent loop, maxIterations={}", config.maxIterations());

        // Reset state
        interrupted.set(false);
        iteration.set(0);
        state = LoopState.IDLE;
        errorHandler.reset();  // Reset error handler for new loop

        TokenUsage totalUsage = new TokenUsage();

        // Add user message to session
        if (prompt != null && !prompt.isEmpty()) {
            session = session.addMessage(Message.user(prompt));
        }

        while (iteration.get() < config.maxIterations()) {
            // Check for interruption
            if (interrupted.get()) {
                state = LoopState.INTERRUPTED;
                logger.info("Loop interrupted at iteration {}", iteration.get());
                return LoopResult.interrupted(session, iteration.get());
            }

            // Build context
            state = LoopState.BUILDING_CONTEXT;
            ContextBuilder.Context context = contextBuilder.build(session);

            // Call LLM with retry
            state = LoopState.CALLING_LLM;
            logger.debug("Calling LLM at iteration {}", iteration.get());

            List<LLMClient.ToolDefinition> tools = convertTools(toolExecutor.listTools());
            LLMResponse response = callLLMWithRetry(context, tools);

            if (response == null) {
                // All retries failed
                state = LoopState.ERROR;
                return LoopResult.error(session, iteration.get(), "LLM call failed after all retries");
            }

            logger.debug("LLM response: stopReason={}, hasToolCalls={}",
                response.stopReason(), response.isToolUse());

            // Update token usage
            totalUsage = totalUsage.add(response.usage());

            // Add assistant message (skip if empty and has tool calls)
            if (response.content() != null && !response.content().isEmpty() || !response.isToolUse()) {
                session = session.addMessage(Message.assistant(response.content()));
            }

            // Check if we need tools
            if (response.isToolUse()) {
                state = LoopState.EXECUTING_TOOLS;

                // Execute tools with timeout
                List<ToolResult> results = executeToolsWithTimeout(response.toolCalls(), session);

                // Add tool results to session
                for (ToolResult result : results) {
                    session = session.addMessage(Message.fromToolResult(result));
                }

                iteration.incrementAndGet();
                continue;
            }

            // Done!
            state = LoopState.COMPLETED;
            session = session.withTokenUsage(totalUsage);

            logger.info("Loop completed at iteration {}, tokens={}",
                iteration.get(), totalUsage.totalTokens());

            return LoopResult.completed(session, response.content(), iteration.get(), totalUsage);
        }

        // Max iterations reached
        state = LoopState.ERROR;
        logger.warn("Max iterations reached: {}", iteration.get());

        // Try to get final response from last assistant message
        String finalResponse = null;
        for (Message msg : session.messages()) {
            if ("assistant".equals(msg.role()) && msg.contentAsString() != null) {
                finalResponse = msg.contentAsString();
            }
        }

        return LoopResult.maxIterations(session, iteration.get());
    }

    /**
     * Call LLM with retry and intelligent error handling.
     *
     * Retry strategy:
     * - Max retries from config.retryOnError()
     * - Uses ErrorHandler for intelligent delay (e.g., Retry-After header)
     * - Fallback: exponential backoff with jitter
     *
     * @return LLMResponse or null if all retries failed
     */
    private LLMResponse callLLMWithRetry(ContextBuilder.Context context, List<LLMClient.ToolDefinition> tools) {
        int maxRetries = config.retryOnError();

        for (int attempt = 0; attempt < maxRetries; attempt++) {
            try {
                return llmClient.call(context.messages(), tools, context.systemPrompt());
            } catch (Exception e) {
                logger.warn("LLM call failed (attempt {}/{}): {}",
                    attempt + 1, maxRetries, e.getMessage());

                if (attempt < maxRetries - 1) {
                    // Build error context
                    ErrorContext errorContext = ErrorContext.builder()
                        .error(e)
                        .iteration(iteration.get())
                        .attempt(attempt + 1)
                        .build();

                    // Get intelligent decision from ErrorHandler
                    ErrorDecision decision = errorHandler.handle(e, errorContext);

                    if (decision.action() == ErrorAction.RETRY) {
                        // Use ErrorHandler's delay or fallback to exponential backoff with jitter
                        double delaySeconds;
                        if (decision.delaySeconds() > 0) {
                            delaySeconds = decision.delaySeconds();
                        } else {
                            // Exponential backoff with jitter (cap at 30s)
                            long baseBackoffMs = Math.min((long) Math.pow(2, attempt) * 1000, 30_000);
                            long jitterMs = random.nextLong(500);
                            delaySeconds = (baseBackoffMs + jitterMs) / 1000.0;
                        }

                        logger.info("Retrying in {}s: {}", String.format("%.1f", delaySeconds), decision.message());

                        try {
                            Thread.sleep((long) (delaySeconds * 1000));
                        } catch (InterruptedException ie) {
                            Thread.currentThread().interrupt();
                            logger.warn("Retry sleep interrupted");
                            break;
                        }
                    } else if (decision.action() == ErrorAction.ABORT) {
                        logger.error("Error handler decided to abort: {}", decision.message());
                        break;
                    }
                    // Other actions (SKIP, COMPRESS_CONTEXT, ESCALATE) not handled in LLM call
                }
            }
        }

        logger.error("LLM call failed after {} attempts", maxRetries);
        return null;
    }

    /**
     * Execute tool calls with timeout protection.
     *
     * Each tool execution is limited by config.timeoutPerTool().
     * If timeout is exceeded, returns a failed ToolResult.
     */
    private List<ToolResult> executeToolsWithTimeout(List<ToolCall> toolCalls, Session session) {
        ToolContext context = ToolContext.builder()
            .sessionId(session.id())
            .workingDirectory(config.workingDirectory())
            .iteration(iteration.get())
            .build();

        List<CompletableFuture<ToolResult>> futures = new ArrayList<>();
        for (ToolCall call : toolCalls) {
            futures.add(executeToolWithTimeout(call, context));
        }

        // Wait for all tools to complete
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

        return futures.stream()
            .map(CompletableFuture::join)
            .toList();
    }

    /**
     * Execute a single tool with timeout.
     */
    private CompletableFuture<ToolResult> executeToolWithTimeout(ToolCall call, ToolContext context) {
        CompletableFuture<ToolResult> future = toolExecutor.execute(call, context);

        return future.orTimeout(config.timeoutPerTool(), TimeUnit.MILLISECONDS)
            .exceptionally(ex -> {
                if (ex.getCause() instanceof TimeoutException) {
                    logger.warn("Tool {} timed out after {}ms", call.name(), config.timeoutPerTool());
                    return ToolResult.error(
                        call.id(),
                        call.name(),
                        "Tool execution timed out after " + config.timeoutPerTool() + "ms"
                    );
                }
                logger.error("Tool {} failed: {}", call.name(), ex.getMessage());
                return ToolResult.error(call.id(), call.name(), ex.getMessage());
            });
    }

    /**
     * Convert internal tools to LLM tool definitions.
     */
    private List<LLMClient.ToolDefinition> convertTools(List<Tool> tools) {
        return tools.stream()
            .map(t -> LLMClient.ToolDefinition.of(t.name(), t.description(), t.inputSchema()))
            .toList();
    }

    /**
     * Interrupt the current loop.
     */
    public void interrupt() {
        interrupted.set(true);
    }

    /**
     * Get current state.
     */
    public LoopState getState() {
        return state;
    }

    /**
     * Get current iteration.
     */
    public int getIteration() {
        return iteration.get();
    }
}